rfcs/0007-filtering.md

```diff
@@ -0,0 +1,211 @@
+            if (!fieldMapping.containsKey(field.getName())) {
+                PropertyUtils.setProperty(objToTrim, field.getName(), null);
+                continue;
+            }
+
+            // Return the entire field in the results
+            SubfieldSelection selection = fieldMapping.get(field.getName());
+            if (selection.getPickMe() == WhichFields.ALL) {
+                continue;
+            }
+
+            // Return a subset of the fields of the object referenced.
+            trimFieldsThrowException(objToTrim, field, selection);
+        }
+    }
+
+    private static void trimFieldsThrowException(Object objToTrim, Field field, SubfieldSelection selection) {
+        try {
+            trimFields(field.get(objToTrim), selection.getExplictFields());
+        } catch (IllegalAccessException iae) {
+            throw new RuntimeException(iae);
+        }
+    }
+``` 
+
+### Proposed Solution: Unresolved Questions
+[Proposed Solution: Unresolved Questions]: #proposed-solution-unresolved-questions
+
+How much will this save us?
+ 
+Proposed profiling is invoking BFD with at least 10,000 EOB Requests by Patient ID (ExplanationOfBenefit?patient=) and to measure
+the aggregate time of completion of all calls.  In other words, sum up the time taken by each call over all calls.
+
+This test will be baselined on current implementation and then ran again after the solution is deployed.
+
+
+### Proposed Solution: Drawbacks
+[Proposed Solution: Drawbacks]: #proposed-solution-drawbacks
+
+This solution incurs nominal amount of extra processing to traverse the EOB object and null out fields that do not
+need to be returned.
+
+This solution adds AB2D specific code within the BFD source base.  Any new required fields results in changes to BFD
+code.  The expecation is that this is rare and involves minimal effort.
+
+### Proposed Solution: Notable Alternatives
+[Proposed Solution: Notable Alternatives]: #proposed-solution-notable-alternatives
+
+GraphQL is a notable alternative where the fields are explicitly requested in the invocation.  Since GraphQL
+supports introspection, the client can actively determine the available fields.  A hybrid approach using
+*registered queries* requires queries to be specified in advance so that they can be locked down by a production server.
+
+The downside is that this would be a significant amount of work for BFD to adopt a different API style while continuing
+to support the REST calls.
+
+An alternative for using a library like PropertyUtils would be JXPath, but initial benchmarks showed that PropertyUtils
+was consistently faster.
+
+## Future Possibilities
+[Future Possibilities]: #future-possibilities
+
+There could be dynamic projections on the database side as well. Right now there are 194 columns
+in the beneficiary table, and we certainly do not need all of them, so retrieving only the needed columns
+would reduce the load on the database.
```

Comment:
I agree in general, but just as an FYI: right now, BFD's performance is **heavily** dominated by JSON serialization -- over 80% of response time is spent on that. So DB performance is a long ways down the list, relatively speaking. (I'll also be curious to see if these filtering changes improve serialization performance!)

Reply:
+1

Reply:
`Additional goals for the harness are to assess how much can concurrency increase.  Currently, AB2D limits EOB
retrievals to 32 concurrent requests.  The plans are to run cooperative experiments to establish a maximum concurrent limit.`

Have you considered frameworks like netty? I've seen netty handle 20k concurrent med device emulator connections running SSL WebSockets. I'm sure you could blow by the 32 concurrent without even coming close to taxing the system.